import gensim
import gensim.corpora as corpora
from gensim.corpora import Dictionary
from gensim.models.coherencemodel import CoherenceModel
from gensim.models.ldamodel import LdaModel

from pprint import pprint

import spacy

import pickle
import re 
import pyLDAvis
import pyLDAvis.gensim

import matplotlib.pyplot as plt 
import pandas as pd
___

import numpy
numpy.random.bit_generator = numpy.random._bit_generator

### fully uninstalled, reinstalled numpy, restarted kernel, finally 
import gensim
from gensim import corpora
#### about 18 cells down runs, creates:
###### Dictionary(54130 unique tokens: ['0', '0frac', '0mathbf', '0ttfrac', '1']...)

import numpy
numpy.random.bit_generator = numpy.random._bit_generator

import numpy as np
rng = np.random.default_rng(12345)
print(rng)
Generator(PCG64)
rfloat = rng.random()
rfloat
0.22733602246716966
type(rfloat)

random.Generator.bit_generator

with help from

https://shichaoji.com/tag/topic-modeling-python-lda-visualization-gensim-pyldavis-nltk/

In [6]:
from helper import *

ModuleNotFoundError: No module named 'helper'

In [7]:
import warnings

In [8]:
warnings.filterwarnings('ignore')

In [9]:
import nltk
nltk.download('stopwords')
nltk.download('wordnet')

C:\Users\Owner\Anaconda3\lib\site-packages\sklearn\feature_extraction\image.py:167: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  dtype=np.int):
C:\Users\Owner\Anaconda3\lib\site-packages\sklearn\linear_model\least_angle.py:30: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.o

True

In [10]:
from nltk.corpus import stopwords 
from nltk.stem.wordnet import WordNetLemmatizer
import string

stopwords = set(stopwords.words('english'))
punctuation = set(string.punctuation) 
lemmatize = WordNetLemmatizer()

def cleaning(article):
    one = " ".join([i for i in article.lower().split() if i not in stopwords])
    two = "".join(i for i in one if i not in punctuation)
    three = " ".join(lemmatize.lemmatize(i) for i in two.split())
    return three

In [11]:
import pandas as pd
import numpy as np
import re
import json
import glob

In [12]:
folder = (glob.glob('../data/english_topic_full_content/*'))

In [22]:
article_list = []
for file in folder:
    with open(file, 'r') as f:
        article = json.load(f)
    article_list.append(article)

In [25]:
articles = pd.DataFrame(article_list).replace(r'^\s*$', np.nan, regex = True)
articles.info()

#ignore error - dataframe still runs

TypeError: Cannot interpret '<attribute 'dtype' of 'numpy.generic' objects>' as a data type

In [26]:
articles = articles.dropna()

In [27]:
articles.head(200)

topic english_title  \
0    Acceleration  Acceleration   
1      acceptance    Acceptance   
4       Aerospace     Aerospace   
5    Agricultural   Agriculture   
6     Agriculture   Agriculture   
..            ...           ...   
262      Shopping      Shopping   
264        Social        Social   
265       Society       Society   
266     Sociology     Sociology   
267      software      Software   

                                               english  
0    In mechanics, acceleration is the rate of chan...  
1    Acceptance in human psychology is a person's a...  
4    Aerospace is a term used to collectively refer...  
5    Agriculture is the science, art and practice o...  
6    Agriculture is the science, art and practice o...  
..                                                 ...  
262  Shopping is an activity in which a customer br...  
264  Social organisms, including humans, live colle...  
265  A society is a group of individuals involved i...  
266  Sociology is the study of society, human socia...  
267  Software is a collection of instructions and d...  

[200 rows x 3 columns]

In [17]:
text = articles.applymap(cleaning)['english']
text_list = [i.split() for i in text]
len(text_list)

243

In [18]:
text_list[0]

['mechanic',
 'acceleration',
 'rate',
 'change',
 'velocity',
 'object',
 'respect',
 'time',
 'acceleration',
 'vector',
 'quantity',
 'in',
 'magnitude',
 'direction',
 'orientation',
 'object',
 'acceleration',
 'given',
 'orientation',
 'net',
 'force',
 'acting',
 'object',
 'magnitude',
 'object',
 'acceleration',
 'described',
 'newton',
 'second',
 'law',
 'combined',
 'effect',
 'two',
 'cause',
 'net',
 'balance',
 'external',
 'force',
 'acting',
 'onto',
 'object',
 '—',
 'magnitude',
 'directly',
 'proportional',
 'net',
 'resulting',
 'force',
 'object',
 'mass',
 'depending',
 'material',
 'made',
 '—',
 'magnitude',
 'inversely',
 'proportional',
 'object',
 'massthe',
 'si',
 'unit',
 'acceleration',
 'metre',
 'per',
 'second',
 'squared',
 'm⋅s−2',
 '2',
 'displaystyle',
 'tfrac',
 'operatorname',
 'm',
 'operatorname',
 's',
 '2',
 'example',
 'vehicle',
 'start',
 'standstill',
 'zero',
 'velocity',
 'inertial',
 'frame',
 'reference',
 'travel',
 'straight',
 'li

In [19]:
from time import time
import logging

logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.INFO,
                   filename='running.log',filemode='w')

In [20]:
# Importing Gensim
import gensim
from gensim import corpora

# Creating the term dictionary of our courpus, where every unique term is assigned an index. dictionary = corpora.Dictionary(doc_clean)
dictionary = corpora.Dictionary(text_list)
dictionary.save('dictionary.dict')
print(dictionary)

Dictionary(54130 unique tokens: ['0', '0frac', '0mathbf', '0ttfrac', '1']...)


In [30]:
# Converting list of documents (corpus) into Document Term Matrix using dictionary prepared above.
doc_term_matrix = [dictionary.doc2bow(doc) for doc in text_list]
corpora.MmCorpus.serialize('corpus.mm', doc_term_matrix)

len(doc_term_matrix)
print(doc_term_matrix[100])

[(8, 1), (21, 1), (23, 1), (27, 3), (29, 10), (30, 2), (31, 1), (32, 2), (42, 5), (47, 4), (50, 2), (51, 3), (52, 2), (63, 3), (64, 2), (70, 1), (75, 2), (76, 1), (88, 1), (103, 2), (108, 1), (109, 2), (110, 1), (123, 1), (131, 2), (145, 1), (147, 6), (153, 1), (155, 2), (160, 1), (164, 5), (168, 1), (177, 1), (180, 1), (191, 1), (193, 1), (199, 2), (213, 1), (215, 5), (223, 1), (229, 3), (233, 1), (237, 8), (254, 2), (261, 3), (268, 3), (269, 1), (272, 9), (274, 6), (281, 1), (286, 4), (288, 3), (292, 2), (300, 2), (303, 2), (316, 1), (328, 3), (329, 1), (335, 1), (343, 1), (348, 1), (349, 1), (350, 3), (352, 2), (353, 2), (374, 8), (380, 1), (383, 17), (394, 3), (399, 3), (401, 1), (408, 4), (409, 1), (419, 3), (429, 2), (468, 2), (477, 1), (482, 3), (492, 1), (494, 6), (496, 1), (505, 2), (518, 2), (520, 1), (522, 1), (532, 4), (546, 2), (556, 2), (561, 1), (564, 2), (565, 1), (571, 1), (572, 1), (573, 1), (575, 1), (576, 2), (584, 1), (585, 2), (608, 3), (609, 2), (610, 1), (625, 1

In [32]:
start = time()
# Creating the object for LDA model using gensim library
Lda = gensim.models.ldamodel.LdaModel

# Running and Trainign LDA model on the document term matrix.
ldamodel = Lda(doc_term_matrix, num_topics=10, id2word = dictionary, passes=50)
print('used: {:.2f}s'.format(time()-start))

used: 220.90s


In [33]:
print(ldamodel.print_topics(num_topics=2, num_words=4))

[(2, '0.010*"theory" + 0.005*"one" + 0.005*"education" + 0.005*"memory"'), (0, '0.007*"therapy" + 0.007*"library" + 0.007*"health" + 0.006*"medical"')]


In [35]:
for i in ldamodel.print_topics(): 
    for j in i: print(j)

0
0.007*"therapy" + 0.007*"library" + 0.007*"health" + 0.006*"medical" + 0.006*"medicine" + 0.006*"woman" + 0.005*"physical" + 0.005*"also" + 0.005*"massage" + 0.005*"disease"
1
0.008*"law" + 0.006*"religion" + 0.005*"history" + 0.005*"also" + 0.004*"wikipedia" + 0.004*"car" + 0.004*"world" + 0.004*"one" + 0.004*"century" + 0.003*"rome"
2
0.010*"theory" + 0.005*"one" + 0.005*"education" + 0.005*"memory" + 0.005*"also" + 0.005*"learning" + 0.004*"human" + 0.004*"mind" + 0.004*"study" + 0.004*"logic"
3
0.007*"cell" + 0.007*"plant" + 0.006*"organism" + 0.006*"system" + 0.006*"energy" + 0.005*"specie" + 0.005*"animal" + 0.004*"life" + 0.004*"one" + 0.004*"form"
4
0.013*"language" + 0.009*"war" + 0.008*"time" + 0.006*"child" + 0.005*"military" + 0.004*"used" + 0.004*"may" + 0.004*"also" + 0.004*"use" + 0.004*"mining"
5
0.011*"map" + 0.010*"gender" + 0.008*"political" + 0.006*"social" + 0.005*"state" + 0.004*"system" + 0.004*"also" + 0.004*"woman" + 0.004*"society" + 0.004*"survey"
6
0.008*"

In [36]:
ldamodel.save('topic.model')

In [37]:
from gensim.models import LdaModel
loading = LdaModel.load('topic.model')

In [38]:
print(loading.print_topics(num_topics=2, num_words=4))

[(8, '0.006*"human" + 0.006*"science" + 0.006*"social" + 0.005*"study"'), (7, '0.006*"world" + 0.005*"year" + 0.005*"population" + 0.005*"china"')]


In [39]:
def pre_new(doc):
    one = cleaning(doc).split()
    two = dictionary.doc2bow(one)
    return two

In [40]:
pre_new('new article that to be classified by trained model!')

[(261, 1), (495, 1), (740, 1), (4711, 1), (8806, 1)]

In [41]:
belong = loading[(pre_new('new article that to be classified by trained model!'))]
belong

[(0, 0.2035756),
 (1, 0.6630084),
 (2, 0.016680181),
 (3, 0.01667828),
 (4, 0.016675446),
 (5, 0.016675444),
 (6, 0.01667364),
 (7, 0.016675701),
 (8, 0.016681153),
 (9, 0.016676111)]

In [42]:
new = pd.DataFrame(belong,columns=['id','prob']).sort_values('prob',ascending=False)
new['topic'] = new['id'].apply(loading.print_topic)
new

id      prob                                              topic
1   1  0.663008  0.008*"law" + 0.006*"religion" + 0.005*"histor...
0   0  0.203576  0.007*"therapy" + 0.007*"library" + 0.007*"hea...
8   8  0.016681  0.006*"human" + 0.006*"science" + 0.006*"socia...
2   2  0.016680  0.010*"theory" + 0.005*"one" + 0.005*"educatio...
3   3  0.016678  0.007*"cell" + 0.007*"plant" + 0.006*"organism...
9   9  0.016676  0.013*"food" + 0.007*"art" + 0.007*"entertainm...
7   7  0.016676  0.006*"world" + 0.005*"year" + 0.005*"populati...
4   4  0.016675  0.013*"language" + 0.009*"war" + 0.008*"time" ...
5   5  0.016675  0.011*"map" + 0.010*"gender" + 0.008*"politica...
6   6  0.016674  0.008*"area" + 0.006*"forensic" + 0.006*"leisu...

In [43]:
new['topic']

1    0.008*"law" + 0.006*"religion" + 0.005*"histor...
0    0.007*"therapy" + 0.007*"library" + 0.007*"hea...
8    0.006*"human" + 0.006*"science" + 0.006*"socia...
2    0.010*"theory" + 0.005*"one" + 0.005*"educatio...
3    0.007*"cell" + 0.007*"plant" + 0.006*"organism...
9    0.013*"food" + 0.007*"art" + 0.007*"entertainm...
7    0.006*"world" + 0.005*"year" + 0.005*"populati...
4    0.013*"language" + 0.009*"war" + 0.008*"time" ...
5    0.011*"map" + 0.010*"gender" + 0.008*"politica...
6    0.008*"area" + 0.006*"forensic" + 0.006*"leisu...
Name: topic, dtype: object

In [66]:
import gensim
from gensim import corpora

In [119]:
ran next cell again

SyntaxError: invalid syntax (<ipython-input-119-7d0cb2b8f431>, line 1)

In [117]:
import pyLDAvis.gensim
import gensim
pyLDAvis.enable_notebook()

###### conda install -c conda-forge pyldavis finally did it

In [56]:
import pyLDAvis
import pyLDAvis.gensim_models as gensimvis
pyLDAvis.enable_notebook()

C:\Users\Owner\Anaconda3\lib\site-packages\sklearn\decomposition\online_lda.py:29: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  EPS = np.finfo(np.float).eps


Successfully built sklearn
Installing collected packages: funcy, sklearn, pandas, pyLDAvis
  Found existing installation: pandas 0.25.1
    Uninstalling pandas-0.25.1:
      Successfully uninstalled pandas-0.25.1
ERROR: Could not install packages due to an EnvironmentError: [WinError 5] Access is denied: 'c:\\users\\owner\\anaconda3\\lib\\site-packages\\~andas\\io\\msgpack\\_packer.cp37-win_amd64.pyd'
Consider using the `--user` option or check the permissions.

##### worked anyway?  .. then it worked

In [120]:
d = gensim.corpora.Dictionary.load('dictionary.dict')
c = gensim.corpora.MmCorpus('corpus.mm')
lda = gensim.models.LdaModel.load('topic.model')

###### then, pip install pyLDAvis==2.1.2 or not

In [68]:
import gensim
from gensim import corpora

##### after got to end, came back and ran this again
The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
--> 232         use_numexpr = use_numexpr and _bool_arith_check(op_str, a, b)
    233         if use_numexpr:
    234             # error: "None" not callable

In [124]:
articles = pyLDAvis.gensim.prepare(lda, c, d)
articles

C:\Users\Owner\Anaconda3\lib\site-packages\pandas\core\indexes\base.py:459: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  True (duplicates aren't allowed).


ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

In [123]:
articles = pyLDAvis.gensim.prepare(optimal_model, corpus, id2word)
articles

NameError: name 'optimal_model' is not defined

In [96]:
pyLDAvis.enable_notebook()
articles = pyLDAvis.gensim.prepare(lda, c, d)
articles

C:\Users\Owner\Anaconda3\lib\site-packages\pandas\core\indexes\base.py:459: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  True (duplicates aren't allowed).


ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

In [71]:
from pyLDAvis import gensim

###### worked anyway??

In [101]:
pyLDAvis.save_html(articles,'vis.html')

In [102]:
display(HTML(filename = 'vis.html'))

NameError: name 'HTML' is not defined

In [103]:
import codecs
import webbrowser

In [104]:
webbrowser.open('vis.html')

True

trying with pyLDAvis.enable_ntebook(local=True) worked for me - some forum